### PR TITLE
Set blinding=yes property on some algorithm implementations

### LINKED ISSUE #14654: Blinding property
From #14616.

Add a `blinding` property and set this to _yes_ for our relevant implementations.  We will not supply `blinding=no` implementations.

### PR DESCRIPTION
**Pull request by t8m**
Fixes #14654

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by romen** [comment]
Why is this property even visible?

**Comment by t8m** [comment]
> Why is this property even visible?

This was discussed within OTC. It is a const and read-only replacement of RSA_FLAG_NO_BLINDING flag.

**Comment by romen** [comment]
Can't that flag be just ignored? I don't see a reason why a user would ever pick a non-blinding implementation: my understanding is that that flag exists mostly for compatibility with engines and code written before 1.0, but transitioning into providers could make that detail irrelevant for applications.

**Comment by t8m** [comment]
That flag will be ignored. Adding the property is IMO harmless. Feel free to bring this up during the OTC meeting though.

**Comment by richsalz** [comment]
Adding things, especially those that will almost never be used, is not harmless.  It adds to the conceptual complexity and things *everyone* has to understand.  I support @romen's view.

File: doc/man7/provider-asym_cipher.pod [hunk 1]
```diff
@@ -164,6 +164,21 @@ The length of the decrypted data should be written to I<*outlen>.
 If I<out> is NULL then the maximum length of the decrypted data should be
 written to I<*outlen>.
 
+=head2 Properties
+
+The following property is set by some of the OpenSSL asymmetric cipher
+algorithms. Third party implementations can set additional properties not
```
Comment (paulidale) [inline doc/man7/provider-asym_cipher.pod:170]: I'd leave out the second sentence: _Third party implementations can set additional properties not mentioned here._

File: doc/man7/provider-signature.pod [hunk 2]
```diff
@@ -325,6 +325,20 @@ initialised verification context is passed in the I<ctx> parameter. The data to
 verified is in I<tbs> which should be I<tbslen> bytes long. The signature to be
 verified is in I<sig> which is I<siglen> bytes long.
 
+=head2 Properties
+
+The following property is set by some of the OpenSSL signature algorithms.
+Third party implementations can set additional properties not mentioned here.
```
Comment (paulidale) [inline doc/man7/provider-signature.pod:331]: I'd delete this line.

File: doc/man7/provider-asym_cipher.pod (hunk 1 above)
Comment (t8m) [reply doc/man7/provider-asym_cipher.pod]: OK

File: doc/man7/provider-signature.pod (hunk 2 above)
Comment (t8m) [reply doc/man7/provider-signature.pod]: OK

**Comment by t8m** [comment]
@paulidale fixup pushed

**Review by paulidale — APPROVED** [review]
Looks good.

**Comment by paulidale** [comment]
One thing I did wonder about was adding the constant-time property name at the same time.  It's one we know we want in due course.

**Comment by romen** [comment]
I wanted to discuss this but couldn't at today meeting.

I still fail to see why there should be a knob to advertise/select algorithms based on things like "it supports blinding", "it is resistant against timing attacks", etc.

To me they are equivalent to letting a user query/select a crypto algorithm based on a "secure=yes" property.

- Does it add one more place where we risk to encourage applications (especially in combination with external providers) to be insecure by default (if they forgot to query with `consttime=yes, blinding=yes`or if a future version of openssl introduces `foosecure=yes`)? 
- what user/dev problem is this mechanism trying to solve? (The legacy blinding_rsa flag should be an internal thing of the past for legacy paths that have to deal with engines, right?) Failing to see one, why are we adding this as a visibile property that could ossify into a worst practice/API bad design?

**Comment by t8m** [comment]
There was a lengthy discussion on this topic during the OTC meeting today. Unfortunately I cannot fully reproduce the discussion here.

>     * Does it add one more place where we risk to encourage applications (especially in combination with external providers) to be insecure by default (if they forgot to query with `consttime=yes, blinding=yes`or if a future version of openssl introduces `foosecure=yes`)?

I do not see how omitting `blinding=yes` from a query makes the implementation selected to be insecure by default. It just gives you the opportunity to select the secure implementation if there is insecure one present as well. The presence of the property does not make the problem of selecting an implementation worse.

One of the things we discussed was the potential `consttime=yes` property. If we implemented it properly for current implementations that are not fully constant time it would allow you to prefer a better third party implementation that is const time. And yes, of course better solution would be to _fix_ our implementation to be const time rather than omit the consttime=yes property, but that is a few orders of magnitude harder to do at least for some of the currently non-const time algorithms. In other contexts the consttime implementation might be seriously penalizing in terms of the performance and in context where const time is not required it is quite limiting that you cannot provide const and non-const time implementation to select from accordingly.

Security is not a boolean value.

**Comment by t8m** [comment]
> One thing I did wonder about was adding the constant-time property name at the same time. It's one we know we want in due course.

Heh, that is much harder endeavor at least for not just adding the property but properly assigning it to individual algorithm implementations.

**Comment by romen** [comment]
> There was a lengthy discussion on this topic during the OTC meeting today. Unfortunately I cannot fully reproduce the discussion here.
> 
> > ```
> > * Does it add one more place where we risk to encourage applications (especially in combination with external providers) to be insecure by default (if they forgot to query with `consttime=yes, blinding=yes`or if a future version of openssl introduces `foosecure=yes`)?
> > ```
> 
> I do not see how omitting `blinding=yes` from a query makes the implementation selected to be insecure by default. It just gives you the opportunity to select the secure implementation if there is insecure one present as well. The presence of the property does not make the problem of selecting an implementation worse.
> 
> One of the things we discussed was the potential `consttime=yes` property. If we implemented it properly for current implementations that are not fully constant time it would allow you to prefer a better third party implementation that is const time. And yes, of course better solution would be to _fix_ our implementation to be const time rather than omit the consttime=yes property, but that is a few orders of magnitude harder to do at least for some of the currently non-const time algorithms. In other contexts the consttime implementation might be seriously penalizing in terms of the performance and in context where const time is not required it is quite limiting that you cannot provide const and non-const time implementation to select from accordingly.
> 
> Security is not a boolean value.

Exactly this: I believe it shouldn't be left to applications to set up propqueries to pick secure alternatives over insecure ones. If an external implementation offers better guarantees than the builtin modules there are already ways to prefer the external provider over the default provider. That is why I fail to see a benefit to having these properties here, while I see drawbacks from usability, design and slippery slopes.

**Comment by t8m** [comment]
@romen feel free to put OTC hold on this and propose a vote. Otherwise it will be merged.

**Comment by richsalz** [comment]
@romen, I still support your view (https://github.com/openssl/openssl/pull/14759#issuecomment-811914521)

**Comment by paulidale** [comment]
Those at the OTC meeting yesterday were in a consensus that having a property like this was a good thing.  The question was should it be a 3.0 thing or not.

The gist of the argument was that blinding is a property of an algorithm: either an implementation does blinding or it doesn't.  There are lots of implementations out there that don't do blinding.  How does a user distinguish between them when fetching?  Property queries are our decision making process for implementation selection.  That all of our implementations will gain the blinding property is moot, other implementations won't necessarily have it.  For a user who really wants blinding, and I don't see why anyone wouldn't, they can force the selection of an implementation that includes it.

const-time is more complicated but essentially the same thing: const-time prevents a number of attack avenues but imposes a performance overhead.  Being able to ensure the implementation you select is constant time is useful to a user.  Likewise, being about to select one that isn't (which is presumably) faster is also useful.

This was all nutted out way back at the Brisbane design meetings, it's been planned the entire time.

**Comment by paulidale** [comment]
@romen, at @t8m wrote: put an OTC hold on this for a vote if you object strongly enough.
The call yesterday was quorate but we didn't call a vote.

**Comment by romen** [comment]
I will formulate a vote proposal as soon as I can, sorry I couldn't take part into the discussion on the OTC call to clear this up sooner.

**Comment by romen** [comment]
The vote proposal is published: https://www.mail-archive.com/openssl-project@openssl.org/msg02407.html

Please provide feedback about it on the mailing list!

File: providers/defltprov.c [hunk 3]
```diff
@@ -369,15 +369,15 @@ static const OSSL_ALGORITHM deflt_rands[] = {
 
 static const OSSL_ALGORITHM deflt_signature[] = {
 #ifndef OPENSSL_NO_DSA
-    { "DSA:dsaEncryption", "provider=default", ossl_dsa_signature_functions },
+    { "DSA:dsaEncryption", "provider=default,blinding=yes", ossl_dsa_signature_functions },
```
Comment (slontis) [inline providers/defltprov.c:372]: put space after,

**Comment by romen** [comment]
The vote has now been closed, rejecting this PR:

~~~
----------------
topic: Reject PR#14759
Proposed by Nicola Tuveri
Public: yes
opened: 2021-04-20
closed: 2021-05-04
accepted:  yes  (for: 3, against: 2, abstained: 4, not voted: 2)
  Matt       [ 0]
  Mark       [  ]
  Pauli      [-0]
  Viktor     [  ]
  Tim        [-1]
  Richard    [ 0]
  Shane      [+1]
  Tomas      [-1]
  Kurt       [+1]
  Matthias   [ 0]
  Nicola     [+1]

~~~

**Pull request closed without merging**